### DISCUSSION TITLE
Create shader from naga::Module

### DISCUSSION THREAD

**Discussion by wrnrlr** [gfx-rs/wgpu#1868] [Q&A]
I am looking into building a procedural macro in Rust that outputs a WSGL shader and was wondering how a `naga::Module` can be used directly instead of `wgpu::ShaderSource`. I found that in `wgpu::Context::device_create_shader_module` this is already being done based of a `str` through `wgc::pipeline::ShaderModuleSource::Wgsl`.

```
ShaderSource::Wgsl(ref code) => wgc::pipeline::ShaderModuleSource::Wgsl(Borrowed(code))
```

I was thinking of using `naga::Module` as the target of my shader transpiler instead of just concatenating strings with bits of WGSL.

Please tell me if this is not a good idea, and otherwise what the best way is to implement this feature?

**Comment by kvark** [comment]
I'm hesitating to make wgpu-rs accept a naga::Module because that would make `naga` a public dependency. Currently, it's a private dependency. So we can update `naga` without breaking the outside API.

Also, accepting `naga::Module` directly is not going to help the Web target, it will only be available on native.

Finally, Naga's IR is very close WGSL. Conversion to WGSL is very lightweight, and so is parsing WGSL. So you can just convert Naga IR to WGSL and pass it to us, portably.

**Reply by wrnrlr** [reply]
Thats sounds fair, I later relised that `naga::Module` probably doesn't belong in the Wgpu's public interface.
Can I still reuse the structs in `naga::front::wgsl` to represent the target AST of my transpiler, if so how can I turn this tree into a string, or do you thing it is better to write my own version of the wgsl grammar?

**Reply by kvark** [reply]
You can use Naga IR, and you'd use the WGSL backend to turn it into text for feeding into wgpu. See https://github.com/gfx-rs/naga/blob/cb0ad2504ce8bbd868237032f49c10c944b490cb/cli/src/main.rs#L413

**Reply by wrnrlr** [reply]
Thx, I'll do that
